File: datasets/Python/Copilot/Algorithm/hard4.py
from collections import deque

def reverse_factorization(N, A):
    A.sort()
    visited = set()
    q = deque([(1, [1])])
    while q:
        curr, path = q.popleft()
        for a in A:
            next_val = curr * a
            if next_val > N or next_val in visited or N % next_val != 0:
                continue
            visited.add(next_val)
            new_path = path + [next_val]
            if next_val == N:
                return new_path
            q.append((next_val, new_path))
    return [-1]

# Manual Input
N_K = input("Enter N and K (e.g. 12 3): ").strip().split()
N, K = int(N_K[0]), int(N_K[1])
A = list(map(int, input(f"Enter {K} integers for A: ").strip().split()))

result = reverse_factorization(N, A)
print("Answer:", ' '.join(map(str, result)))

File: datasets/C++/Copilot/Algorithm/hard4.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <unordered_set>
using namespace std;

int main() {
    int N, K;
    cout << "Enter N and K: ";
    cin >> N >> K;
    vector<int> A(K);
    cout << "Enter " << K << " integers for A: ";
    for (int i = 0; i < K; ++i) cin >> A[i];
    sort(A.begin(), A.end());

    queue<pair<long long, vector<long long>>> q;
    unordered_set<long long> visited;
    q.push({1, {1}});
    visited.insert(1);

    while (!q.empty()) {
        auto [curr, path] = q.front(); q.pop();
        for (int a : A) {
            long long next_val = curr * a;
            if (next_val > N || N % next_val != 0 || visited.count(next_val)) continue;
            visited.insert(next_val);
            vector<long long> new_path = path;
            new_path.push_back(next_val);
            if (next_val == N) {
                for (auto v : new_path) cout << v << " ";
                cout << endl;
                return 0;
            }
            q.push({next_val, new_path});
        }
    }
    cout << -1 << endl;
    return 0;
}
